### DISCUSSION TITLE
[Idea] SEO improvements for PrestaShop

### DISCUSSION THREAD

**Discussion by adamus007p** [PrestaShop/PrestaShop#32283] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

I would like to report few things to correct regards SEO in Prestashop.

1. There is no canonical for main website
It should be `<link rel="canonical" href="https://www.site.com/">`
any lang main websites should be `<link rel="canonical" href="https://www.site.com/de">` for Germany for example

2. Canonical for dynamin pages. Here is a plage for improvments. For example for categories where there are many pages.
example: category page, page 3 `https://www.site.com/5-example-category?page=3`
Link canonical shoud be to main category page not for `page=3`
It should be: `<link rel="canonical" href="https://www.site.com/5-example-category>`
not `<link rel="canonical" href="https://www.site.com/5-example-category?page=3">` like currenlty it is.
More over all dynamic pages (`?page=3`, filters and so on) should have meta:
`<meta name="robots" content="noindex,follow">`

3. I would consider to keep one URL pattern.
Now categories are with no / in the end of the category, but languges have / in the end: `site.com/de/`

### Steps to reproduce

Check the source code for main page and for categories.

### PrestaShop version(s) where the bug happened

1.7.8.6

**Comment by adamus007p** [comment]
I can make changes, but please advise how to send my changes.

**Comment by AureRita** [comment]
Hi @adamus007p 

Thank you for the reporting Issue.
if I understand correctly, you advise us to make an improvement on the SEO but it is not a bug to reproduce? 
If so, it is a feature and we will consider it as such to treat it. To do this, could you show us an ideal link and the wrong link on the other side.
If not, in the Steps to reproduce section, can you tell us what you have done to have a link like this for your shop so that we can redo it on our shops on our side and thus correct the problem

Thanks you

**Comment by adamus007p** [comment]
Hello for me it is a bug, but it will impove SEO in Prestashop.
I have implement it in my shop and I can share how to correct it.
@AureRita Please advise me how can i update the repo?

**Comment by AureRita** [comment]
If you want to update a repo, 
- go on this link : https://github.com/PrestaShop/PrestaShop/pulls 
- clic on " New pull reset " 

but you need an Approval for this new PR

**Comment by AureRita** [comment]
Did your problem seems like this one ? 

https://github.com/PrestaShop/PrestaShop/issues/13737

**Comment by adamus007p** [comment]
Not really. I have described what should be corrected.

**Comment by matks** [comment]
Hello @adamus007p , you're starting a complex topic 😅 

Every time a SEO question is asked to 6 people we get at least 3 or 4 different answers 😛 (see example https://github.com/PrestaShop/PrestaShop/issues/20677)

Thank you for the suggestion but as we have a lot of people telling us "you should do [...]" "no! you should do [...] for canonicals!" "no! wrong! you should do [...] for canonical urls" we have to compare what you tell with some other SEO knowledge.

So I dont think we can consider what you report as a bug. A bug is a defect, something that we know clearly how it should behave and it does not follow. You're suggesting improvements / feature requests about how SEO is handled

@AureRita this should be submitted to Product team I believe

**Comment by adamus007p** [comment]
Hello @matks I know that it is complex.
My experience in SEO is more then 12years. I had many Prestashop and I see the same problems.
All dynamic pages should be noindex.

What I have done I have change a theme, but maybe it is possible to do it in controller.

`{if isset($category) && preg_match('/(page|order|selected_filters|tag|order|p|amp|back|search_query|n)\=/',$request_uri)}
        <meta name="robots" content="noindex, follow">
    {/if}`

site.com/1-category?page=2, page=3 etc I will be a duplicate content what google do not like.

It is easy and you can implement it.

**Comment by Hlavtox** [comment]
@adamus007p 

1) Homepage definitely should have canoncial. Good catch! 

2) You are wrong. Different pages have different content and must have a different canoncial. Prestashop have this implemented correctly. 

See google documentation - https://developers.google.com/search/docs/advanced/ecommerce/pagination-and-incremental-page-loading

> Give each page a unique URL. For example, include a ?page=n query parameter, as URLs in a paginated sequence are treated as separate pages by Google.
> 
> Don't use the first page of a paginated sequence as the canonical page. Instead, give each page in its own [canonical URL](https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls).

3) Good idea!

**Comment by adamus007p** [comment]
Hello @Hlavtox , Regards 2. there are diferent schools. I am talking what I see from my tests.

Anyway all filters should have `<meta name="robots" content="noindex,follow">` and in my opinion paginations too.

**Comment by kpodemski** [comment]
@adamus007p _in your opinion_ - we decided based on the research we did that it's not a strategy we want to follow

you can always implement these changes on your side 👍

**Comment by adamus007p** [comment]
@kpodemski  Maybe opinion is wrong word. It is called experience. Try find big websites who have good SEO with pagination in the search results ;)

I see what is wrong or missing regards SEO. I have shared it with you. Rest it is up to you. If you have more experience then me please share. Maybe I will learn something new. I am open for a discussion.


I will clarify my previous sentence. All dynamic pages should have tag `<meta name="robots" content="noindex, follow">`. This is besides the canonical differences, which is not so important right now for google.

**Comment by Hlavtox** [comment]
@adamus007p

> I will clarify my previous sentence. All dynamic pages should have tag <meta name="robots" content="noindex, follow">. This is besides the canonical differences, which is not so important right now for google.

This is already done by automatically generated robots.txt file.

Can you please check if homepage is the only page we are missing a canoncial? So we can fix it all at once. ;-)

**Comment by Hlavtox** [comment]
> I regret joining this thread, the best I can do now is to cancel the star, not pay a penny for prestashop's modules and themes again, and no longer recommend PrestaShop to anyone.

bye bye

**Comment by adamus007p** [comment]
> is is already done by automatically generated robots.txt file.

> Can you please check if homepage is the only page we are missing a canonical? So we can fix it all at once. ;-)

In the theory yes, in real no.

Only main page do not have canonical.

**Comment by HartLarsson** [comment]
SEO have directive and best practise given also by Google and other gained from experiences:

https://developers.google.com/search/docs/advanced/ecommerce/pagination-and-incremental-page-loading

**Comment by AureRita** [comment]
Ping @PrestaShop/product-team friendly reminder

**Comment by adamus007p** [comment]
@AureRita missing canonical for the main page it is a bug, or regression when it was in Prestashop 1.6

**Comment by adamus007p** [comment]
@HartLarsson as you see google show only basic things, when you have ecommerce shop you have other filters like page 2 and price, etc, which make another possible variations for links. Google like every company has some budget for SEO for all pages and when see that some shop have a lot of links it will not crawl them all.

For example `?page=2` can be crawled by robot when `order=product.name.asc&page=2` and others make a duplication.


When `?page=2` can be follow other dynamic filters should be `<meta name="robots" content="noindex, nofollow">`.

**Comment by adamus007p** [comment]
I have another idea for Prestashop to introduce that you can set up noindex for certain categories. Like you can setup it in CMS.

For example one of my client has category Types, it mean nothing for google and people, but inside it there are some subcategories and say what it is inside. It is only needed for the shop structure.
The same I have delete it from breadcrumb and from SEO giving it noindex.

It would be nice functions by default in Prestashop.


Implement what can be indexed or in breadcrumbs it would be a nice functionality. Regards categories, products, filters etc.
It could be  default recommended setup, but everyone could change it in easy way to test the SEO and changes in SEO.
What do you think? @AureRita @matks @kpodemski @Hlavtox @HartLarsson

**Comment by Kerami82** [comment]
> Hello for me it is a bug, but it will impove SEO in Prestashop. I have implement it in my shop and I can share how to correct it. @AureRita Please advise me how can i update the repo?

@adamus007p please tell me where I can find your solution for "Canonical for dynamin pages". Im looking for that :(

**Comment by HartLarsson** [comment]
> @HartLarsson as you see google show only basic things, when you have ecommerce shop you have other filters like page 2 and price, etc, which make another possible variations for links. Google like every company has some budget for SEO for all pages and when see that some shop have a lot of links it will not crawl them all.
> 
> For example `?page=2` can be crawled by robot when `order=product.name.asc&page=2` and others make a duplication.
> 
> When `?page=2` can be follow other dynamic filters should be `<meta name="robots" content="noindex, nofollow">`.
 
i did a personal modification in pagination-seo.tpl to keep in canonical only the page information and keep dynamic filters out the canonical url.
